### ISSUE TITLE
Botan 2.1 Release Plan

### ISSUE THREAD

**Issue by randombit** [randombit/botan#866]
Given the current pace of development I think a ~ quarterly release cycle for new minor versions is appropriate. A minor bump should not have too few changes, but also not too much. I'm somewhat arbitrarily setting a target date for 2.1.0 of April 1, with all major changes merged to master by Mar 15th to allow time for testing.

New in master already since 2.0:
- NEON in SIMD_4x32
- ARM feature detection using getauxval
- OS::run_cpu_instruction_probe to allow other feature detection as required
- Solaris and SunCC fixes
- Some better perf data (cycles/byte output, measuring pow_mod directly, etc)

Implemented but not merged:
- SHA instructions for Goldmont and ARMv8 - basically done, waiting on some build changes because both intrinsics require GCC more recent than 4.8.

Things I plan to do, > 80% confidence of completion in time for 2.1
- Adding build-time compile checks for OS features, not assuming ability to run the resulting binary (since that breaks cross-compiles) but simply checking for the ability to compile an executable referencing `mmap` or `getauxval` or whatever we are curious about. Then the static files specifying features can be removed. A previous related project that I will probably reuse some code from is https://github.com/randombit/tinned
- Merging Windows Phone support (#794 #795 #796)
- Supporting ARM feature detection via probe functions (for Windows, iOS, ...)
- ARMv8 PMULL support for GCM (#842)
- Handle implicit CBC IVs (#864)

Stretch goals (would like to do, but probably will not have time for between now and April) - if anyone is interested in taking one of these on, let me know. In no particular order:
- Stiched or combined AES/GCM mode for AES-NI. That is, implementing the entire AES-128/GCM operation using AES-NI, without anything leaving xmm registers during processing. This would probably greatly improve bulk GCM performance for AES-NI systems, maybe important for TLS. This could also be done for ARMv8.
- ChaCha using SIMD_4x32 (already have SSE2 specific code, but this would support VMX and NEON also, and allows a good test if the class wrapper causes any overhead).
- ARMv8 AES support
- POWER8 SHA and AES support
- Investigate running the TPM unit tests in CI against a TPM emulator (https://github.com/PeterHuewe/tpm-emulator)
- Use TLS-Attacker to test error cases of TLS client and server
- Test against BoringSSL's test suite.
- Write a script to measure performance against OpenSSL (probably just calling `botan speed` and `openssl speed` for different algorithms and parsing the outputs) and display or graph the results in some useful way.

Possibly also:
- @cordney mentioned working on supporting TLS server authenticating via a key stored in a TPM, but that it was slow going.

This is not a complete feature list for 2.1 - if someone sends a good patch I'll merge it and I may find myself inspired to make some change I have not thought of yet. But I find it useful to have a sense of what the release is going to look like. I used to just keep notes like these in a text file on my machine but going forward I will create a release ticket that tracks the overall goals and progress of each new feature (minor version) release.

**Comment by vielmetti** [comment]
Thanks @randombit . If you are tracking benchmark results, I have access to interesting ARMv8 hardware (Packet's 2A 96-core Cavium based server) and would be happy to share results.

**Comment by randombit** [comment]
@vielmetti That would be lovely thank you. Post them in #844 so that all notes on the PR are in the same spot.

**Comment by randombit** [comment]
March 16th and I haven't hit any of my target projects so I'm not sure if it is worth releasing 2.1.0 on April 1 or anytime soon after that.

I guess in any case the first step is updating the release notes to capture the full set of changes made on master so far since 2.0, and identify what PRs can likely be ready to merge in the next week or so. Looking especially at #899, #900, #906, and maybe #884.

Some short term preprocessor hackery would enable merging #826 and #844 without requiring the configure changes I have not had time for.

**Comment by cordney** [comment]
Would it instead make sense to release a 2.0.2 at this point? I see at least #913, #912, #902, #891, the expired name constraints test certificates and the renaming of `botan.py` to `botan2.py` would be good reasons to do so (despite all the other fixes).

**Comment by randombit** [comment]
Good reasons all, sold on a new release as previously planned released ~ April 1. I think it will be 2.1.0 regardless, since there are new API features.

**Comment by cordney** [comment]
Great, thanks!

**Comment by randombit** [comment]
OK a little late for April 1st release of 2.1.0 - my plan is now tomorrow April 4th. I think #885, #964 and #965 are all ok for 2.1, just waiting on CI. I also have one more small PR to make, coming shortly. Plus updating the release notes.

**Comment by randombit** [comment]
2.1 was released, more or less on time even, so closing this
